include/util/integer_range.hpp

```diff
@@ -12,35 +13,72 @@ namespace osrm
+// Can be removed after dropping support of Boost < 1.56
+template <typename Integer>
+class integer_iterator : public boost::iterator_facade<integer_iterator<Integer>,
+                                                       Integer,
+                                                       boost::random_access_traversal_tag,
+                                                       Integer>
+{
+    typedef boost::iterator_facade<integer_iterator<Integer>,
+                                   Integer,
+                                   boost::random_access_traversal_tag,
+                                   Integer>
+        base_t;
+
+  public:
+    typedef typename base_t::value_type value_type;
+    typedef typename base_t::difference_type difference_type;
+    typedef typename base_t::reference reference;
+    typedef std::random_access_iterator_tag iterator_category;
+
+    integer_iterator() : m_value() {}
+    explicit integer_iterator(value_type x) : m_value(x) {}
+
+  private:
+    void increment() { ++m_value; }
+    void decrement() { --m_value; }
+    void advance(difference_type offset) { m_value += offset; }
+    bool equal(const integer_iterator &other) const { return m_value == other.m_value; }
+    reference dereference() const { return m_value; }
+
+    difference_type distance_to(const integer_iterator &other) const
+    {
+        return std::is_signed<value_type>::value
+                   ? (other.m_value - m_value)
+                   : (other.m_value >= m_value)
+                         ? static_cast<difference_type>(other.m_value - m_value)
+                         : -static_cast<difference_type>(m_value - other.m_value);
+    }
+
+    friend class ::boost::iterator_core_access;
+    value_type m_value;
+};
+
 // Warning: do not try to replace this with Boost's irange, as it is broken on Boost 1.55:
 //     auto r = boost::irange<unsigned int>(0, 15);
 //     std::cout << r.size() << std::endl;
 // results in -4294967281. Latest Boost versions fix this, but we still support older ones.
 
 template <typename Integer> class range
 {
-  private:
-    const Integer last;
-    Integer iter;
-
   public:
-    range(Integer start, Integer end) noexcept : last(end), iter(start)
-    {
-        BOOST_ASSERT_MSG(start <= end, "backwards counting ranges not suppoted");
-        static_assert(std::is_integral<Integer>::value, "range type must be integral");
-    }
+    typedef integer_iterator<Integer> const_iterator;
+    typedef integer_iterator<Integer> iterator;
 
-    // Iterable functions
-    const range &begin() const noexcept { return *this; }
-    const range &end() const noexcept { return *this; }
-    Integer front() const noexcept { return iter; }
-    Integer back() const noexcept { return last - 1; }
+    range(Integer begin, Integer end) : iter(begin), last(end) {}
+
+    iterator begin() const noexcept { return iter; }
+    iterator end() const noexcept { return last; }
```

Comment:
This part corresponds https://github.com/boostorg/range/blob/develop/include/boost/range/iterator_range_core.hpp#L202-L204 and https://github.com/boostorg/range/blob/develop/include/boost/range/iterator_range_core.hpp#L221-L229 so iterator constness is ignored and i just used boost iterators interface.

I think the method `const` qualifier should be treated as a `begin` call does not change the range, so the second `begin` call will return the same value. An  iterator constness will be relevant in the `dereference` method, but `boost::iterator_facade` defines `reference` as `Integer`, so in an integer range no big difference between const and non-const iterators.